frontend/src/utils/budgetUtil.ts

```diff
@@ -0,0 +1,158 @@
+import { Category, Transaction } from 'src/types/budget';
+
+export interface MonthlyDepositWithdraw {
+  [month: string]: DepositWithdraw;
+}
+
+export interface DepositWithdraw {
+  deposit: number;
+  withdraw: number;
+}
+
+export interface CategoryMap {
+  expense: number;
+  limit: number;
+}
+
+export interface CategoryMaps {
+  [categoryKey: string]: CategoryMap;
+}
+
+export interface CategoryWithdrawMap {
+  [categoryName: string]: number;
+}
+
+export const generateCategoryMaps = (
+  categories: Category[],
+  transactions: Transaction[]
+) => {
+  const result = initializeCategoryMaps(categories);
+  for (const transaction of transactions) {
+    const categoryNames = getCategoryNames(transaction.categoryId, categories);
+    const amount = transaction.withdraw ? Number(transaction.withdraw) : 0;
+    let key = '';
+    for (const categoryName of categoryNames) {
+      key += categoryName;
+      result[key].expense += amount;
+      key += ' / ';
+    }
+  }
+  return result;
+};
+
+export const generateParentCategoryWithdrawMaps = (
+  transactions: Transaction[],
+  categories: Category[]
+): CategoryWithdrawMap => {
+  return transactions.reduce(
+    (acc: CategoryWithdrawMap, transaction: Transaction) => {
+      const amount = transaction.withdraw ? Number(transaction.withdraw) : 0;
+      const categoryName = transaction.categoryId
+        ? getParentCategoryNameById(transaction.categoryId, categories)
```

Comment:
Same as above.
Pass category as argument and you don't need to iterate in `getParentCategoryNameById` function.